scripts/gen_database.py

```diff
@@ -0,0 +1,170 @@
+  "app-persistence"      : "./app"            ,
+  "app-phala"            : "./app"            ,
+  "app-polkadex"         : "./app"            ,
+  "app-polkadot"         : "app"              ,
+  "app-polymesh"         : "./app"            ,
+  "app-reef"             : "./app"            ,
+  "app-secret"           : "./app"            ,
+  "app-stacks"           : "./app"            ,
+  "app-statemine"        : "./app"            ,
+  "app-statemint"        : "./app"            ,
+  "app-thorchain"        : "./app"            ,
+  "app-terra"            : "./app"            ,
+  "app-xxnetwork"        : "./app"            ,
+}
+
+
+def format_database(database: dict) -> str:
+    database_str = json.dumps(database, indent=2, sort_keys=True)
+    # Drop some newlines to compact a bit the data while still
+    # making it readable.
+    database_str = database_str.replace("[\n      ", "[")
+    database_str = database_str.replace("\n    ]", "]")
+    database_str = database_str.replace("\n      ", " ")
+    return database_str
+
+
+def check_database_lint(database_path: Path):
+    ret = 0
+    with open(database_path, 'r') as f:
+        database_str = f.read()
+        database = json.loads(database_str)
+
+    for variant, params in database.items():
+        for param, value in params.items():
+            if param not in APP_LOAD_PARAMS_VALUE_CHECK:
+                print(f"[ERROR] Not allowed '{param}' in variant '{variant}'")
+                ret = -1
+            if not isinstance(value, list):
+                print(f"[ERROR] '{param}' value should be of type list in variant '{variant}'")
+                ret = -1
+
+    # Splitlines put keep line ends and escape it to make them visible.
+    a = [repr(x)[1:-1] for x in database_str.splitlines(keepends=True)]
+    b = [repr(x)[1:-1] for x in format_database(database).splitlines(keepends=True)]
+    res = list(difflib.unified_diff(a, b))
+    if res:
+        ret = -1
+        print("[ERROR] Database is not properly linted, see diff below:")
+        for line in res:
+            print(line)
+
+    if ret != 0:
+        exit(ret)
+
+
+def load_database(database_path: Path):
+    database = {}
+    if database_path.exists():
+        with open(database_path, 'r') as f:
+            database = json.load(f)
+    return database
+
+
+def save_database(database: dict, database_path: Path):
+    database_str = format_database(database)
+    with open(database_path, 'w') as f:
+        f.write(database_str)
+
+
+def gen_app(app_path: Path, database_path: Path):
+    print(f"Generating for {app_path.name}")
+
+    app_build_path = BUILD_PATH_LIST.get(app_path.name, "./")
+    app_full_path = app_path / app_build_path
+
+    # Retrieve database
+    database = load_database(database_path)
+
+    # Retrieve available variants
+    try:
+        variant_param_name, variants = get_app_listvariants(app_full_path)
+    except:
+        print("Skipping generation due to error")
```

Comment:
What's the benefit of silencing the (I guess useful) error information while still highlighting that an error occurred?

Reply:
Allowing the generation for all the apps in a directory which might includes rust apps that don't have listvariant support (nor even Makefile).